csharp/src/Ice/Connection.cs

```diff
@@ -674,2069 +463,1333 @@ public void AsyncRequestCanceled(OutgoingAsyncBase outAsync, System.Exception ex
-            //
-            if (sentCBs != null)
-            {
-                foreach (OutgoingMessage m in sentCBs)
-                {
-                    if (m.InvokeSent)
+                    if (_acmLastActivity > -1)
                     {
-                        Debug.Assert(m.OutAsync != null);
-                        m.OutAsync.InvokeSent();
+                        _acmLastActivity = Time.CurrentMonotonicTimeMillis();
                     }
-                    if (m.ReceivedReply)
+                    if (_connector != null)
                     {
-                        Debug.Assert(m.OutAsync != null);
-                        var outAsync = (OutgoingAsync)m.OutAsync;
-                        if (outAsync.Response(m.IncomingData))
-                        {
-                            outAsync.InvokeResponse();
-                        }
+                        _validated = true;
                     }
+
+                    SetState(State.Active);
                 }
-                ++dispatchedCount;
-            }
 
-            //
-            // Asynchronous replies must be handled outside the thread
-            // synchronization, so that nested calls are possible.
-            //
-            if (info.OutAsync != null)
-            {
-                info.OutAsync.InvokeResponse();
-                ++dispatchedCount;
+                // Start the asynchronous operation from the thread pool to prevent eventually reading
+                // synchronously new frames from this thread.
+                _ = Task.Run(() => RunIO(ReadAsync));
             }
-
-            if (info.HeartbeatCallback != null)
+            catch (Exception ex)
             {
-                try
-                {
-                    info.HeartbeatCallback(this);
-                }
-                catch (Exception ex)
+                lock (_mutex)
                 {
-                    _logger.Error("connection callback exception:\n" + ex + '\n' + _desc);
+                    SetState(State.Closed, ex);
                 }
-                ++dispatchedCount;
-            }
-
-            //
-            // Method invocation must be done outside the thread synchronization, so that nested
-            // calls are possible.
-            //
-            if (info.InvokeNum > 0)
-            {
-                ValueTask vt = InvokeAllAsync(info.Data, info.InvokeNum, info.RequestId, info.Compress,
-                    info.Adapter!);
-
-                // TODO: do something with the value task
-
-                //
-                // Don't increase dispatchedCount, the dispatch count is
-                // decreased when the incoming reply is sent.
-                //
+                throw;
             }
 
-            //
-            // Decrease dispatch count.
-            //
-            if (dispatchedCount > 0)
+            // Helper to await task with timeout
+            static async ValueTask AwaitWithTimeout(Task task, int timeout)
```

Comment:
Passing `ValueTask` doesn't work because `InitializeAsync` returns a `ValueTask` but `Read/Write` return `ValueTask<int>`. I fixed AwaitWithTimeout to check if the task was completed and avoid the creating the Delay task in this case.

Reply:
Doesn't this defeat a bit the advantage of using Value task in the first place, calling AsTask can cause the extra allocation

Reply:
Yes, it would in a performance critical context. Here the calls are performed for the connection establishment which I don't really see as performance critical.

I guess we could change InitializeAsync to return a Task but then it wouldn't be very consistent with other transport methods which are all returning a ValueTask.